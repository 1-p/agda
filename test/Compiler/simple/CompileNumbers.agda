{-# OPTIONS -v treeless.opt:20 #-}
module _ where

open import Common.Prelude
open import Common.Integer

-- Should compile to
--   λ a b → case a of
--              "neg" → 0 - b
--              _     → b
match-on-lit : String → Integer → Integer
match-on-lit "neg" x with x
...                     | pos (suc n) = negsuc n
...                     | pos 0       = pos 0
...                     | negsuc n    = pos (suc n)
match-on-lit _     x = x

-- This doesn't compile as nicely, since the match on "neg"
-- ends up between the match on the int and the nat (not sure why).
match-on-lit₂ : String → Integer → Integer
match-on-lit₂ "neg" (pos (suc n)) = negsuc n
match-on-lit₂ "neg" (negsuc n)    = pos (suc n)
match-on-lit₂ _     x             = x

-- Should compile to a flat case
nested-match : Integer → String
nested-match (pos 0) = "zero"
nested-match (pos 1) = "one"
nested-match (pos (suc (suc n))) = "lots"
nested-match (negsuc 0) = "minus one"
nested-match (negsuc 1) = "minus two"
nested-match (negsuc (suc (suc n))) = "minus lots"

printInt : Integer → IO Unit
printInt x = putStrLn (intToString x)

main : IO Unit
main = printInt (match-on-lit "neg" (pos 42))
    ,, printInt (match-on-lit₂ "neg" (pos 42))
    ,, putStrLn (nested-match (negsuc 5))
