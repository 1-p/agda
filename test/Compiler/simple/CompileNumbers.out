EXECUTED_PROGRAM

ret > ExitSuccess
out > CompileNumbers.nested-match =
out >   λ a →
out >     case a of
out >       0 → "zero"
out >       1 → "one"
out >       _ | a >= 2 → "lots"
out >       -1 → "minus one"
out >       -2 → "minus two"
out >       _ → "minus lots"
out > CompileNumbers.match-on-lit =
out >   λ a b →
out >     case a of
out >       "neg" → 0 - b
out >       _ → b
out > CompileNumbers.printInt =
out >   λ a → Common.IO.putStrLn (Common.String.intToString a)
out > CompileNumbers.match-on-lit₂ =
out >   λ a b →
out >     case b of
out >       _ | b >= 0 →
out >         case a of
out >           "neg" →
out >             case b of
out >               _ | b >= 1 → 0 - b
out >               _ → b
out >           _ → b
out >       _ | b < 0 →
out >         case a of
out >           "neg" → 0 - b
out >           _ → b
out >       _ → b
out > CompileNumbers.main =
out >   Common.IO.then
out >     () () _ _
out >     (CompileNumbers.printInt (CompileNumbers.match-on-lit "neg" 42))
out >     (Common.IO.then
out >        () () _ _
out >        (CompileNumbers.printInt (CompileNumbers.match-on-lit₂ "neg" 42))
out >        (Common.IO.putStrLn (CompileNumbers.nested-match -6)))
out > -42
out > -42
out > minus lots
out >
